### PR TITLE
migrate bcc based system monitor to libbpf and cilium/ebpf

### LINKED ISSUE #38: ContainerMonitor: use libbpf instead of bcc
bcc requires kernel headers and it seems unstable when we test KubeArmor on several environments (with different kernel versions). Thus, we're planning to change the base of the container monitor from bcc to libbpf.

### PR DESCRIPTION
**Pull request by daemon1024**
**Why?**
- BCC compiles programs at runtime, which has heavy CPU and memory resource usage.
- eliminate `libbcc` dependency on the host machine
- smaller main image size, since compilation is decoupled from the main container.
- Making KubeArmor binary portable provided no CGO dependency

Follow Up #399 

Fixes #38 

TODO
- [x] Test systemd on centos/rhel
- [x] Handle GKE Headers
- [x] Update Docs

### CODE REVIEW

File: KubeArmor/BPF/system_monitor.c [hunk 1]
```diff
@@ -15,6 +15,12 @@
 #define asm_inline asm
 #endif
```
Comment (daemon1024) [inline KubeArmor/BPF/system_monitor.c:16]: Despite adding this in accordance with https://github.com/iovisor/bcc/issues/2546

Still getting the same errors when using the default version of clang provided in the `ubuntu-18.04` image in gh action, 
image
Ref https://github.com/kubearmor/KubeArmor/runs/6305329041?check_suite_focus=true

Any thoughts on why would this be happening?

It resolves if we upgrade the version of the clang which has the asm inline construct available, but still..

Comment (nam-jaehyun) [reply KubeArmor/BPF/system_monitor.c:16]: what clang version does Ubuntu-18.04 use? what is the minimum version of clang to make KubeArmor work?

Comment (nam-jaehyun) [reply KubeArmor/BPF/system_monitor.c:16]: what clang version does Ubuntu-18.04 use? what is the minimum version of clang to make KubeArmor work?

Comment (daemon1024) [reply KubeArmor/BPF/system_monitor.c:16]: > what clang version does Ubuntu-18.04 use? 

clang 9 in the GH Action env
clang 6 in the Vagrant env

> what is the minimum version of clang to make KubeArmor work?

It's not that straight forward
clang 6 works fine with kernel version 4.x.x
But with 5.x.x we need atleast clang 12

File: Dockerfile [hunk 2]
```diff
@@ -1,12 +1,26 @@
 # SPDX-License-Identifier: Apache-2.0
 # Copyright 2021 Authors of KubeArmor
 
+### Make compiler image
```
Comment (nam-jaehyun) [inline Dockerfile:4]: not compiling the monitor in runtime?

File: Dockerfile [hunk 3]
```diff
@@ -1,12 +1,26 @@
...
+
+FROM alpine:3.15 as kubearmor-init
+
+RUN apk --no-cache update && \
+    apk --no-cache add bash git && \
+    apk --no-cache add llvm clang make gcc
+
+COPY ./KubeArmor/BPF /KubeArmor/BPF/
+COPY ./KubeArmor/build/compile.sh /KubeArmor/compile.sh
+
```
Comment (nam-jaehyun) [inline Dockerfile:15]: remove one line here

File: KubeArmor/build/compile.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2021 Authors of KubeArmor
+
+cd /KubeArmor/BPF
+make
+cp *.bpf.o /opt/kubearmor/BPF/
```
Comment (nam-jaehyun) [inline KubeArmor/build/compile.sh]: add a new line

**Comment by nam-jaehyun** [comment]
why do we have ci-test-runtime.yml and ci-test-systemd.yml?
why didn't ci-test run in this PR?

File: KubeArmor/build/entrypoint.sh [hunk 5]
```diff
@@ -15,23 +15,20 @@ DEBUG=0
...
-        ARMOR_OPTIONS=${@:2}
-        ;;
-    *)
-        ;;
+"-DEBUG")
+    DEBUG=1
+    ARMOR_OPTIONS=${@:2}
+    ;;
+*) ;;
+
```
Comment (nam-jaehyun) [inline KubeArmor/build/entrypoint.sh]: remove an empty line

**Comment by nam-jaehyun** [comment]
can you rebase and fix the conflicting files?

**Comment by nam-jaehyun** [comment]
if I pull your repo and run 'make run' in a vagrant setup, does KubeArmor work?

**Comment by daemon1024** [comment]
CI wasn't running due to conflicts, Fixed the conflicts it would start running now.

I added systemd and runtime tests to test compilation and working of KubeArmor in both systemd package and with the K8s environment using init containers.

> if I pull your repo and run 'make run' in a vagrant setup, does KubeArmor work?

Yup it would work in your current vagrant setup as well, would need to install llvm-toolchain which can be done using
```
wget https://apt.llvm.org/llvm.sh
chmod +x llvm.sh
sudo ./llvm.sh 12
for tool in "clang" "llc" "llvm-strip"; do
    sudo rm -f /usr/bin/$tool
    sudo ln -s /usr/bin/$tool-12 /usr/bin/$tool
done
```
Also BPF compilation would be needed.
`make -C BPF/`

**Comment by nam-jaehyun** [comment]
@daemon1024 can you confirm if this update works in the following envs?
- [ ] vagrant up
- [ ] vagrant up NETNEXT=1
- [ ] vagrant up OS=centos

When I tested this PR, I got some errors.
```
vagrant@kubearmor-dev:~/KubeArmor/KubeArmor/BPF$ make
make -C /home/vagrant/KubeArmor/KubeArmor/BPF/libbpf/src
make[1]: *** /home/vagrant/KubeArmor/KubeArmor/BPF/libbpf/src: No such file or directory.  Stop.
Makefile:55: recipe for target 'chkdir' failed
make: *** [chkdir] Error 2
```

```
vagrant@kubearmor-dev:~/KubeArmor/KubeArmor$ make run
...
2022-05-29 19:35:36.072466	ERROR	Failed to initialize BPF (bpf module is nil open /home/vagrant/KubeArmor/KubeArmor/BPF/system_monitor.bpf.o: no such file or directory)
github.com/kubearmor/KubeArmor/KubeArmor/log.Errf
	/home/vagrant/KubeArmor/KubeArmor/log/logger.go:102
github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).InitSystemMonitor
	/home/vagrant/KubeArmor/KubeArmor/core/kubeArmor.go:215
github.com/kubearmor/KubeArmor/KubeArmor/core.KubeArmor
	/home/vagrant/KubeArmor/KubeArmor/core/kubeArmor.go:407
main.main
	/home/vagrant/KubeArmor/KubeArmor/main.go:44
runtime.main
	/usr/local/go/src/runtime/proc.go:250
2022-05-29 19:35:36.073502	ERROR	Failed to initialize KubeArmor Monitor
github.com/kubearmor/KubeArmor/KubeArmor/log.Err
	/home/vagrant/KubeArmor/KubeArmor/log/logger.go:97
github.com/kubearmor/KubeArmor/KubeArmor/feeder.(*Feeder).Err
	/home/vagrant/KubeArmor/KubeArmor/feeder/feeder.go:460
github.com/kubearmor/KubeArmor/KubeArmor/core.KubeArmor
	/home/vagrant/KubeArmor/KubeArmor/core/kubeArmor.go:408
main.main
	/home/vagrant/KubeArmor/KubeArmor/main.go:44
runtime.main
	/usr/local/go/src/runtime/proc.go:250
```

**Comment by nam-jaehyun** [comment]
@daemon1024 can you confirm if this update works in the following envs?
- [ ] vagrant up
- [ ] vagrant up NETNEXT=1
- [ ] vagrant up OS=centos

When I tested this PR, I got some errors.
```
vagrant@kubearmor-dev:~/KubeArmor/KubeArmor/BPF$ make
make -C /home/vagrant/KubeArmor/KubeArmor/BPF/libbpf/src
make[1]: *** /home/vagrant/KubeArmor/KubeArmor/BPF/libbpf/src: No such file or directory.  Stop.
Makefile:55: recipe for target 'chkdir' failed
make: *** [chkdir] Error 2
```

```
vagrant@kubearmor-dev:~/KubeArmor/KubeArmor$ make run
...
2022-05-29 19:35:36.072466	ERROR	Failed to initialize BPF (bpf module is nil open /home/vagrant/KubeArmor/KubeArmor/BPF/system_monitor.bpf.o: no such file or directory)
github.com/kubearmor/KubeArmor/KubeArmor/log.Errf
	/home/vagrant/KubeArmor/KubeArmor/log/logger.go:102
github.com/kubearmor/KubeArmor/KubeArmor/core.(*KubeArmorDaemon).InitSystemMonitor
	/home/vagrant/KubeArmor/KubeArmor/core/kubeArmor.go:215
github.com/kubearmor/KubeArmor/KubeArmor/core.KubeArmor
	/home/vagrant/KubeArmor/KubeArmor/core/kubeArmor.go:407
main.main
	/home/vagrant/KubeArmor/KubeArmor/main.go:44
runtime.main
	/usr/local/go/src/runtime/proc.go:250
2022-05-29 19:35:36.073502	ERROR	Failed to initialize KubeArmor Monitor
github.com/kubearmor/KubeArmor/KubeArmor/log.Err
	/home/vagrant/KubeArmor/KubeArmor/log/logger.go:97
github.com/kubearmor/KubeArmor/KubeArmor/feeder.(*Feeder).Err
	/home/vagrant/KubeArmor/KubeArmor/feeder/feeder.go:460
github.com/kubearmor/KubeArmor/KubeArmor/core.KubeArmor
	/home/vagrant/KubeArmor/KubeArmor/core/kubeArmor.go:408
main.main
	/home/vagrant/KubeArmor/KubeArmor/main.go:44
runtime.main
	/usr/local/go/src/runtime/proc.go:250
```

**Comment by daemon1024** [comment]
@nam-jaehyun Can you please run `git submodule update --init --recursive` and try again? 

We would need to ask folks to clone KubeArmor with --recursive flag so as the libbpf submodule is also cloned. 

I will try with the mentioned envs ( I haven't yet tried with CentOS) again and report back.

**Comment by daemon1024** [comment]
> @daemon1024 can you confirm if this update works in the following envs?
> 
>     * [ ]  vagrant up
> 
>     * [ ]  vagrant up NETNEXT=1
> 
>     * [ ]  vagrant up OS=centos
> 

@nam-jaehyun I tested on all 3 envs and KubeArmor should run with a single `make run` and no other pre-requisites.

**Comment by nyrahul** [comment]
@daemon1024 , can you please fix the conflicts? Lets this reviewed/merged asap. thanks

File: KubeArmor/BPF/include/stdarg.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
+/*===---- stdarg.h - Variable argument handling ----------------------------===
```
Comment (nyrahul) [inline KubeArmor/BPF/include/stdarg.h]: duplicating standard headers! Why was this necessary?

File: KubeArmor/BPF/system_monitor.c [hunk 7]
```diff
@@ -1089,67 +1220,80 @@ TRACEPOINT_PROBE(syscalls, sys_exit_openat)
...
 }
 
-int trace_ret_listen(struct pt_regs *ctx)
+SEC("kretprobe/__x64_sys_listen")
+int kretprobe__listen(struct pt_regs *ctx)
 {
     return trace_ret_generic(_SYS_LISTEN, ctx, ARG_TYPE0(INT_T)|ARG_TYPE1(INT_T));
 }
+
+char LICENSE[] SEC("license") = "GPL";
```
Comment (nyrahul) [inline KubeArmor/BPF/system_monitor.c]:
```suggestion
char LICENSE[] SEC("license") = "Dual BSD/GPL";
```

File: KubeArmor/build/kubearmor-test-containerd.yaml [hunk 8]
```diff
@@ -37,13 +37,50 @@ spec:
...
+          - mountPath: /sys/fs/bpf
+            name: sys-fs-bpf-path
+          - mountPath: /sys/kernel/security
+            name: sys-kernel-security-path
+          - mountPath: /sys/kernel/debug
+            name: sys-kernel-debug-path
+          - mountPath: /media/root/etc/os-release
+            name: os-release-path
+            readOnly: true
+          - mountPath: /usr/src
```
Comment (nyrahul) [inline KubeArmor/build/kubearmor-test-containerd.yaml:70]: we will face problems with such `mountPath` .. gke COS latest rapid releases do not have `/usr/src` folder.
Ref: #648 

Is this yaml only for test purpose?

File: KubeArmor/monitor/systemMonitor.go [hunk 9]
```diff
@@ -169,7 +175,7 @@ func NewSystemMonitor(node *tp.Node, logger *fd.Feeder, containers *map[string]t
 
 	mon.Status = true
 	mon.UptimeTimeStamp = kl.GetUptimeTimestamp()
-	mon.HostByteOrder = bcc.GetHostByteOrder()
+	mon.HostByteOrder = binary.LittleEndian
```
Comment (nyrahul) [inline KubeArmor/monitor/systemMonitor.go:175]: can we not hardcode the endianness?

**Review by nyrahul — CHANGES_REQUESTED** [review]
Brilliant piece of work!

Few comments other than specified inline.
I believe we will have issues with the mountPaths that are newly added ... 
Can we get this tested on GKE COS Regular/Rapid release channel images and on AWS Bottlerocket as well?

File: KubeArmor/build/kubearmor-test-containerd.yaml (hunk 8 above)
Comment (daemon1024) [reply KubeArmor/build/kubearmor-test-containerd.yaml:70]: These YAML are only for test purposes. We will have to modify `deployments/get` with the new stuff for end user deployments and kArmor.

Adding that to TODO.

File: KubeArmor/monitor/systemMonitor.go (hunk 9 above)
Comment (daemon1024) [reply KubeArmor/monitor/systemMonitor.go:175]: BCC used unsafe calls to determine the byte order
Ref https://github.com/iovisor/gobpf/blob/2289761f1e2092a7416cd5fd55a218802d997cb6/bcc/perf.go#L110

I would like to defer adding this till #614 and prefer using a better way like [conditional builds](https://groups.google.com/g/golang-nuts/c/3GEzwKfRRQw) to determine this. WDYT?

Comment (nyrahul) [reply KubeArmor/monitor/systemMonitor.go:175]: sounds good. I believe ARM wont work anyways unless this gets fixed.

File: KubeArmor/BPF/include/stdarg.h (hunk 6 above)
Comment (daemon1024) [reply KubeArmor/BPF/include/stdarg.h]: I am really not sure now why we needed it 🙃 Removed it

File: KubeArmor/BPF/Makefile [hunk 10]
```diff
@@ -0,0 +1,86 @@
+ifeq (,$(KRNDIR))
+KRNVER = $(shell uname -r)
+ifneq (,$(wildcard /etc/redhat-release))
+	KRNDIR = /usr/src/$(KRNVER)
+else
+	KRNDIR = /usr/src/linux-headers-$(KRNVER)
+endif
```
Comment (daemon1024) [inline KubeArmor/BPF/Makefile]: Noting will need to handle GKE usr path here as well, because we are using the Makefile to compile in container.

Comment (daemon1024) [reply KubeArmor/BPF/Makefile]: Handled in entrypoint for init container (`compile.sh`)

**Comment by daemon1024** [comment]
Tested on
- GKE Regular Channel
- GKE Rapid Release Channel
- BottleRocket EKS

**Comment by daemon1024** [comment]
Fixed compilation on BottleRocket EKS, The object files were shipping with the docker container causing non native bpf code to run resulting in no telemetry data.

**Review by nyrahul — APPROVED** [review]
LGTM 👍🏼 👍🏼

File: .gitignore [hunk 11]
```diff
@@ -17,6 +17,9 @@ contribution/vagrant/*.log
 *.rpm
 .coverprofile
 
+#Object files
+*.o
```
Comment (DelusionalOptimist) [inline .gitignore]: @daemon1024 We should also remove object files present [here](https://github.com/kubearmor/KubeArmor/tree/main/KubeArmor/enforcer/bpflsm) from the tree, right?

Comment (daemon1024) [reply .gitignore]: Those are needed though, since they cannot be generated at runtime in GH Action (due to lack of BPFLSM support) . Thanks for noticing, I will update the ignore to be `BPF/*.o`

Comment (daemon1024) [reply .gitignore]: The object files in enforcer/bpflsm follow CO-RE ( Compile Once Run Everywhere ) hence it's fine for them to be in the tree, this is not the case for monitor code. Hence the need to clean them out.

Comment (DelusionalOptimist) [reply .gitignore]: Gotcha :+1:

**Comment by DelusionalOptimist** [comment]
Lots of tests now hehe :stuck_out_tongue:

**Review by DelusionalOptimist — APPROVED** [review]
LGTM :rocket:
